### PR TITLE
feat: support addTask and startTask for k8 RP [DET-3416, DET-3419]

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
This PR supports addTask and startTask for Kubernetes. It includes the initial implementation of the Kubernetes RP actor, the pods actor, and the pod actor. So far these actors take task spec and translate it to a pod spec amd submit the pod spec to k8.

## Test Plan
Tested manually by running on a k8 cluster. Adding CI will be part of M4. 

## Commentary
One thing that I wish we could do better is not using scheduler/agentState to represent the Kubernetes resources in the scheduler. However, refactoring this will be a significant amount of work that we will plan to complete as future tech debt work.

### CODE REVIEW

File: master/internal/kubernetes/pods.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		fmt.Sprintf("pod-%s", msg.Spec.TaskID),
+		newPod(
+			p.cluster, p.clientSet, p.namespace, p.masterIP,
+			p.masterPort, msg.Spec, msg.Slots, msg.Rank,
+			p.podInterface, p.configMapInterface,
+		),
+	)
+
+	if !ok {
+		return errors.Errorf(fmt.Sprintf("pod actor %s already exits", ref.Address().String()))
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply master/internal/kubernetes/pods.go]: done

File: master/internal/scheduler/kubernetes_resource_provider.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+			container:   container,
+			clusterID:   k.clusterID,
+			harnessPath: k.harnessPath,
+
+			taskContainerDefaults: k.taskContainerDefaults,
+		})
+}
+
+func (k *kubernetesResourceProvider) getOrCreateGroup(
+	handler *actor.Ref,
```
Comment (dzhu) [inline master/internal/scheduler/kubernetes_resource_provider.go:191]: non-blocking: Let's keep the context first everywhere it's passed as an argument.

File: master/internal/scheduler/kubernetes_resource_provider.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+	if task.SlotsNeeded() > 1 {
+		if k.slotsPerNode == 0 {
+			ctx.Log().WithField("task ID", task.ID).Error(
+				"set slots_per_node > 0 to schedule tasks with slots")
+			return
+		}
+
+		if task.SlotsNeeded()%k.slotsPerNode != 0 {
+			ctx.Log().WithField("task ID", task.ID).Error(
+				"task number of slots is not schedulable on the configured slots_per_node")
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply master/internal/scheduler/kubernetes_resource_provider.go]: done

File: master/internal/scheduler/kubernetes_resource_provider.go (hunk 2 above)
Comment (aaron276h) [reply master/internal/scheduler/kubernetes_resource_provider.go]: fixed

File: master/pkg/tasks/ports.go [hunk 4]
```diff
@@ -9,8 +9,11 @@ import (
 	"github.com/determined-ai/determined/master/pkg/device"
 )
 
-const localRendezvousPort = 1734
-const localRendezvousPortOffset = 16
+// LocalRendezvousPort is the start of the range of ports used for rendezvous by tasks.
+const LocalRendezvousPort = 1734
+
+// LocalRendezvousPortOffset is the offset for rendezvous ports.
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply master/pkg/tasks/ports.go]: updated it

File: master/pkg/tasks/task.go [hunk 5]
```diff
@@ -18,16 +18,17 @@ import (
 )
 
 const (
-	containerWorkDir  = "/run/determined/workdir"
+	// ContainerWorkDir is working directory for containers.
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply master/pkg/tasks/task.go]: updated

File: master/pkg/tasks/task.go [hunk 6]
```diff
@@ -58,19 +59,46 @@ func ToContainerSpec(t TaskSpec) container.Spec {
...
 	user := ""
-	if cmd.AgentUserGroup != nil {
-		user = fmt.Sprintf("%d:%d", cmd.AgentUserGroup.UID, cmd.AgentUserGroup.GID)
+	if agentUserGroup != nil {
+		user = fmt.Sprintf("%d:%d", agentUserGroup.UID, agentUserGroup.GID)
+	}
+	return user
+}
+
+// ConfigureCommandEnvVars configures environment variables for cmd tasks.
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply master/pkg/tasks/task.go]: yep, updated these

File: master/pkg/tasks/task.go [hunk 7]
```diff
@@ -58,19 +59,46 @@ func ToContainerSpec(t TaskSpec) container.Spec {
...
+func ConfigureCommandEnvVars(t TaskSpec) map[string]string {
+	envVarsMap := defaultEnvVars()
+	envVarsMap["DET_TASK_ID"] = t.TaskID
+	return envVarsMap
+}
+
+// ConfigureCommandArchives returns the additional files for a c as an archive.
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply master/pkg/tasks/task.go]: fixed

File: master/pkg/tasks/task.go [hunk 8]
```diff
@@ -230,19 +272,14 @@ func startContainer(t TaskSpec) container.Spec {
...
-		user = fmt.Sprintf("%d:%d", gcc.AgentUserGroup.UID, gcc.AgentUserGroup.GID)
-	}
-	deviceType := device.CPU
-	if len(t.Devices) > 0 {
-		deviceType = t.Devices[0].Type
-	}
-	envVars := defaultEnvVars()
-	envVars = append(envVars, gcc.ExperimentConfig.Environment.EnvironmentVariables.For(deviceType)...)
-	mounts := toDockerMounts(gcc.ExperimentConfig.BindMounts)
+// ConfigureGCEnvVars returns environment variables for gc.
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply master/pkg/tasks/task.go]: updated

**Pull request merged**